iml-agent/src/rpm.rs

```diff
@@ -1,25 +1,85 @@
 use regex::Regex;
 
+use std::fmt;
 use std::process::Output;
+use std::result::Result as StdResult;
 
 use crate::{agent_error::ImlAgentError, cmd::cmd_output};
 
-fn parse(output: Output) -> Result<bool, ImlAgentError> {
+enum PackageState {
+    Installed(Version),
+    NotInstalled,
+}
+
+#[derive(Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Version(String);
+
+impl fmt::Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+pub type Result = StdResult<Version, Error>;
+
+#[derive(Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub enum Error {
+    PackageNotInstalled,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            Self::PackageNotInstalled => "Package not installed",
+        };
+        write!(f, "{}", s)
+    }
+}
+
+async fn parse(output: Output) -> StdResult<PackageState, ImlAgentError> {
     if output.status.success() {
-        Ok(true)
+        // In case there's syntax error in query format, exit code of `rpm` is 0,
+        // but there's no data and an error is on stderr
+        if output.stderr.len() > 0 {
+            Err(ImlAgentError::CmdOutputError(output))
+        } else {
+            Ok(PackageState::Installed(Version(
+                String::from_utf8(output.stdout).unwrap(),
+            )))
+        }
     } else {
         let stdout = output.stdout.clone();
         let re = Regex::new(r"^package .*? is not installed\n$").unwrap();
         let s = String::from_utf8(stdout)?;
         if re.is_match(&s) {
-            Ok(false)
+            Ok(PackageState::NotInstalled)
         } else {
             Err(ImlAgentError::CmdOutputError(output))
         }
     }
 }
 
-pub(crate) async fn installed(package_name: &str) -> Result<bool, ImlAgentError> {
-    let output = cmd_output("rpm", vec!["--query", package_name]).await?;
-    parse(output)
+async fn run_rpm(package_name: &str) -> StdResult<Output, ImlAgentError> {
+    cmd_output(
+        "rpm",
+        vec!["--query", "--queryformat", "%{VERSION}", package_name],
+    )
+    .await
+}
+
+pub(crate) async fn installed(package_name: &str) -> StdResult<bool, ImlAgentError> {
+    let output = run_rpm(package_name).await?;
+    parse(output).await.map(|r| match r {
+        PackageState::Installed(_) => true,
+        PackageState::NotInstalled => false,
+    })
+}
+
+pub(crate) async fn version(package_name: &str) -> StdResult<Result, ImlAgentError> {
+    let output = run_rpm(package_name).await?;
+    match parse(output).await {
+        Ok(PackageState::Installed(v)) => Ok(Result::Ok(v)),
```

Comment:
Here I think that `Option` would be better  aesthetically too, to avoid nested `Result`s.